### PR TITLE
Restricting endpoints

### PR DESCRIPTION
**Pull request by cataraul**

### CODE REVIEW

File: TweetBook/Authorization/WorksForCompanyHandler.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, WorksForCompanyRequirement requirement)
+        {
+            var userEmailAddress = context.User?.FindFirstValue(ClaimTypes.Email) ?? string.Empty;
+
+            if (userEmailAddress.EndsWith(requirement.DomainName))
+            {
+                context.Succeed(requirement);
+                return Task.CompletedTask;
+            }
+
```
Comment (AlinCuc) [inline TweetBook/Authorization/WorksForCompanyHandler.cs]: remove this empty line

**Review by MariaCrucin — APPROVED** [review]

File: TweetBook/Contract/V1/ApiRoutes.cs [hunk 2]
```diff
@@ -28,5 +28,16 @@ public static class Identity
...
             public const string Refresh = Base + "/identity/refresh";
         }
+
+        public static class Tags
+        {
+            public const string GetAll = Base + "/tags";
+
+            public const string Create = Base + "/tags";
+
+            public const string Get = Base + "/tags";
```
Comment (tatarflavia) [inline TweetBook/Contract/V1/ApiRoutes.cs]: I see that GetAll, Create and Get are all equal variables, could you eliminate the duplication?

File: TweetBook/Contract/V1/Requests/CreatePostRequest.cs [hunk 3]
```diff
@@ -3,5 +3,6 @@
     public class CreatePostRequest
     {
         public string Name { get; set; }
```
Comment (tatarflavia) [inline TweetBook/Contract/V1/Requests/CreatePostRequest.cs:5]: Leave a live between the methods for readability.

File: TweetBook/Contract/V1/Requests/CreateTagRequest.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace TweetBook.Contract.V1.Requests
+{
+    public class CreateTagRequest
+    {
+        public string tagName { get; set; }
```
Comment (tatarflavia) [inline TweetBook/Contract/V1/Requests/CreateTagRequest.cs]: In C# it is considered good practice to have all the properties and methods names start with a capital letter.

File: TweetBook/Contract/V1/Responses/PostResponse.cs [hunk 5]
```diff
@@ -3,5 +3,7 @@
     public class PostResponse
     {
         public Guid Id { get; set; }
+        public string Name { get; set; }
```
Comment (tatarflavia) [inline TweetBook/Contract/V1/Responses/PostResponse.cs:7]: Please leave lines between properties for readability.

File: TweetBook/Controllers/V1/PostsController.cs [hunk 6]
```diff
@@ -35,13 +43,23 @@ public async Task<IActionResult> Update([FromRoute] Guid postId,[FromBody] Updat
...
+            {
+                TagName = tagName,
+                PostId = post.Id,
+            }).ToList();
 
-           var updated = await _postService.UpdatePostAsync(post);
-
-            if(updated)
-                return Ok(post);
+            if (!await _postService.UpdatePostAsync(post))
```
Comment (tatarflavia) [inline TweetBook/Controllers/V1/PostsController.cs:52]: Erase the space between the if and the first paranthesis

File: TweetBook/Data/DataContext.cs [hunk 7]
```diff
@@ -11,5 +11,16 @@ public DataContext(DbContextOptions<DataContext> options) : base(options)
...
         public DbSet<RefreshToken> RefreshTokens { get; set; }
+        public DbSet<Tag> Tags { get; set; }
+        public DbSet<PostTag> PostTags { get; set; }
+
+        protected override void OnModelCreating(ModelBuilder builder)
+        {
+            base.OnModelCreating(builder);
+
+            builder.Entity<PostTag>()
+                .Ignore(xx => xx.Post)
```
Comment (tatarflavia) [inline TweetBook/Data/DataContext.cs]: Why did you add this line?

File: TweetBook/Data/DataContext.cs [hunk 8]
```diff
@@ -11,5 +11,16 @@ public DataContext(DbContextOptions<DataContext> options) : base(options)
...
+                .HasKey(x => new { x.PostId, x.TagName });
```
Comment (tatarflavia) [inline TweetBook/Data/DataContext.cs]: Here this x is a PostTag, but x is not a very explicit name for PostTag, is it? 
Please try to use descriptive variables everywhere.

File: TweetBook/Domain/Post.cs [hunk 9]
```diff
@@ -17,5 +17,7 @@ public class Post
 
         [ForeignKey(nameof(UserId))]
         public IdentityUser User { get; set; }
+
+        public List<PostTag> Tags { get; set; }
```
Comment (tatarflavia) [inline TweetBook/Domain/Post.cs]: I think you can remember my comment about using IList such that you can afterwards cast it to whatever list type you want(array, list etc). Another used collection is IEnumerable.

File: TweetBook/Domain/Tag.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace TweetBook.Domain
+{
+    public class Tag
+    {
+        [Key]
+        public string Name { get; set; }
+        public string CreatorId { get; set; }
+
```
Comment (tatarflavia) [inline TweetBook/Domain/Tag.cs:13]: Leave lines between all the properties please

File: TweetBook/Domain/PostTag.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿using System.ComponentModel.DataAnnotations.Schema;
+
+namespace TweetBook.Domain
+{
+    public class PostTag
+    {
+       [ForeignKey(nameof(TagName))]
+       public  virtual Tag Tag { get; set; }
```
Comment (tatarflavia) [inline TweetBook/Domain/PostTag.cs:9]: Why are we only declaring the Tag as a foreign key? Isn't the Post also a foreign key?

File: TweetBook/Domain/PostTag.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+       public string TagName { get; set; }
+       public  virtual Post Post { get; set; }
```
Comment (tatarflavia) [inline TweetBook/Domain/PostTag.cs:14]: Leave lines between properties please

File: TweetBook/Services/IPostService.cs [hunk 13]
```diff
@@ -3,8 +3,8 @@
 namespace TweetBook.Services
 {
     public interface IPostService
-    {
-       public Task<List<Post>> GetPostsAsync();
+    { 
+       public Task<List<Post>> GetAllAsync();
```
Comment (tatarflavia) [inline TweetBook/Services/IPostService.cs]: Please see my comment above about IList

File: TweetBook/Services/TagService.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            }
+
+            _dataContext.Tags.Remove(tagToDelete);
+            var numDeleted = await _dataContext.SaveChangesAsync();
+
+            return numDeleted > 0;
+        }
+        public async Task<Tag> GetAsync(string name)
+        {
+            return await _dataContext.Tags.FirstOrDefaultAsync(t => t.Name == name);
```
Comment (tatarflavia) [inline TweetBook/Services/TagService.cs]: Here t is a Tag, but please choose a more descriptive name for the variable

File: TweetBook/Services/TagService.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        public async Task<bool> CreateAsync(Tag tag)
+        {
+            _dataContext.Tags.Add(tag);
+            var created = await _dataContext.SaveChangesAsync();
+            return created > 0;
```
Comment (tatarflavia) [inline TweetBook/Services/TagService.cs:42]: Leave a blank line before return for consistency in all the methods please

File: TweetBook/Data/DataContext.cs (hunk 7 above)
Comment (cataraul) [reply TweetBook/Data/DataContext.cs]: It adds a primary key, but i added the attribute now.

File: TweetBook/Services/IPostService.cs [hunk 16]
```diff
@@ -4,7 +4,7 @@ namespace TweetBook.Services
 {
     public interface IPostService
     { 
-       public Task<List<Post>> GetAllAsync();
+       public Task<IEnumerable<Post>> GetAllAsync();
```
Comment (tatarflavia) [inline TweetBook/Services/IPostService.cs]: I only mentioned IEnumerable because it is used on projects and for your knowledge. 
You use IEnumerable when you want to loop through the items in a collection. IList is when you want to add, remove, and access the list contents out of order.
So please change to IList

File: TweetBook/Data/DataContext.cs [hunk 17]
```diff
@@ -14,13 +14,6 @@ public DataContext(DbContextOptions<DataContext> options) : base(options)
...
 
-        protected override void OnModelCreating(ModelBuilder builder)
-        {
-            base.OnModelCreating(builder);
-
-            builder.Entity<PostTag>()
-                .Ignore(xx => xx.Post)
-                .HasKey(x => new { x.PostId, x.TagName });
-        }
+
```
Comment (tatarflavia) [inline TweetBook/Data/DataContext.cs]: Here remember the talk on Skype

File: TweetBook/Domain/PostTag.cs [hunk 18]
```diff
@@ -1,13 +1,19 @@
...
+using System.ComponentModel.DataAnnotations.Schema;
 
 namespace TweetBook.Domain
 {
     public class PostTag
     {
        [ForeignKey(nameof(TagName))]
        public  virtual Tag Tag { get; set; }
+
+       [Key]
```
Comment (tatarflavia) [inline TweetBook/Domain/PostTag.cs]: Here as well remember the talk on Skype

File: TweetBook/Services/PostService.cs [hunk 19]
```diff
@@ -12,7 +12,7 @@ public PostService(DataContext dataContext)
         {
             _dataContext = dataContext;
         }
-        public async Task<List<Post>> GetAllAsync()
+        public async Task<IEnumerable<Post>> GetAllAsync()
```
Comment (tatarflavia) [inline TweetBook/Services/PostService.cs]: Here see my above comment

**Review by tatarflavia — APPROVED** [review]
